src/loader/bgw_launcher.c

```diff
@@ -312,78 +334,134 @@ populate_database_htab(HTAB *db_htab)
+{
+	Assert(entry->state == DISABLED);
+	scheduler_modify_state(entry, ENABLED);
+}
+
+static void
+scheduler_state_trans_enabled_to_allocated(DbHashEntry *entry)
+{
+	Assert(entry->state == ENABLED);
+	/* Reserve a spot for this scheduler with BGW counter */
+	if (!bgw_total_workers_increment())
 	{
-		ereport(LOG, (errcode(ERRCODE_CONFIGURATION_LIMIT_EXCEEDED),
-					  errmsg("total databases = %ld TimescaleDB background worker limit %d, so no schedulers allocated", hash_get_num_entries(db_htab), guc_max_background_workers),
-					  errhint("You may start background workers manually by using the  _timescaledb_internal.start_background_workers() function in each database you would like to have a scheduler worker in.")));
+		report_bgw_limit_exceeded(entry);
 		return;
 	}
+	scheduler_modify_state(entry, ALLOCATED);
+}
 
-	/*
-	 * All schedulers have been correctly accounted for, so go ahead and
-	 * insert all into the actual hash table.
-	 */
-	foreach(lc, db_oids)
-		(void) db_hash_entry_create(db_htab, lfirst_oid(lc));
+static void
+scheduler_state_trans_started_to_allocated(DbHashEntry *entry)
+{
+	pid_t		worker_pid;
 
-	/*
-	 * This commit is at the end of this function instead of after heap_close
-	 * above because the db_oids list is allocated on this txn's memory
-	 * context. We cannot free this context until we are done using the list.
-	 */
-	CommitTransactionCommand();
+	Assert(entry->state == STARTED);
+	Assert(get_background_worker_pid(entry->db_scheduler_handle, &worker_pid) == BGWH_STOPPED);
+	scheduler_modify_state(entry, ALLOCATED);
+}
+
+static void
+scheduler_state_trans_allocated_to_start(DbHashEntry *entry)
+{
+	pid_t		worker_pid;
+	bool		worker_registered;
+
+	Assert(entry->state == ALLOCATED);
+
+	worker_registered = register_entrypoint_for_db(entry->db_oid, entry->vxid, &entry->db_scheduler_handle);
+	if (!worker_registered)
+	{
+		report_error_on_worker_register_failure(entry);
+		return;
+	}
+	wait_for_background_worker_startup(entry->db_scheduler_handle, &worker_pid);
+	SetInvalidVirtualTransactionId(entry->vxid);
+	scheduler_modify_state(entry, STARTED);
+}
+
+static void
+scheduler_state_trans_enabled_to_disabled(DbHashEntry *entry)
+{
+	Assert(entry->state == ENABLED);
+	scheduler_modify_state(entry, DISABLED);
+}
+
+static void
+scheduler_state_trans_allocated_to_disabled(DbHashEntry *entry)
+{
+	Assert(entry->state == ALLOCATED);
+	bgw_total_workers_decrement();
+	scheduler_modify_state(entry, DISABLED);
+}
+
+static void
+scheduler_state_trans_started_to_disabled(DbHashEntry *entry)
+{
+	pid_t		worker_pid;
+
+	Assert(entry->state == STARTED);
+	Assert(get_background_worker_pid(entry->db_scheduler_handle, &worker_pid) == BGWH_STOPPED);
+	bgw_total_workers_decrement();
+	scheduler_modify_state(entry, DISABLED);
 }
 
-/* Scan our hash table of dbs and register a worker for each */
+
 static void
-start_db_schedulers(HTAB *db_htab)
+scheduler_state_advance(DbHashEntry *entry)
```

Comment:
Seems inconsistent that this is called `scheduler_state_advance` when all other (sub)functions are called `scheduler_state_trans_...`.

Probably better to also call this `scheduler_state_trans` or scheduler_state_transition`. 

Also related to my comment about the advance_failures variable.

Reply:
renamed to `scheduler_state_trans_automatic`